### PR TITLE
Close #1263 Add Listener/Dispatcher interfaces fornode creation/parsing action

### LINKED ISSUE #1263: Identify and Implement Syntactic and Semantic Observers Design Patterns. (A02.T01)
***Purpose***

The purpose here is to implement Syntactical node creation listeners, and Parser build action dispatchers, so that Syntax directed strategies can be used, applied and elaborated.

# ProgramClassBuilderNodeListener class #

This class is a NodeListener instance over a IProgramClassBuilder interface imlplementation, witch is able to listen to Systactical nodes creation and parsing action performing.

class_type_cobol_1_1_compiler_1_1_cup_parser_1_1_node_builder_1_1_program_class_builder_node_listener__inherit__graph

# ProgramClassBuilderNodeDispatcher class #

This class is a NodeDispatcher instance over a IProgramClassBuilder interface imlplementation, witch is able to dispatch Syntactical nodes created and parsing actions performed, to registered NodeListener instances and ProgramClassBuilderNodeListener instances.

class_type_cobol_1_1_compiler_1_1_cup_parser_1_1_node_builder_1_1_program_class_builder_node_dispatcher__inherit__graph

# Other classes descriptions #

| **Class** | **Description** |
| :------- | :---------- |
| NodeListener | An interface that describes the contract of any objet able to listen to Syntactical node creation. |
| IProgramClassBuilder | An interface that describes the contract of any objet able to handle parsing actions for program class builing. |
| NodeDispatcher | It is the base class of any object which can dispatch node creation actions to registered NodeListener instances. |

### PR DESCRIPTION
**Pull request by smedilol**
Close #1263
I've extracted the commits from branch 1260_DataflowDomain and I fixed conflicts with develop.
I propose to merge this apart from PR #1309

### CODE REVIEW

File: TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs [hunk 1]
```diff
@@ -821,66 +891,77 @@ public virtual void EndExecStatement()
...
         }
 
         public virtual void OnExitProgramStatement(ExitProgramStatement stmt)
         {
             Enter(new ExitProgram(stmt), stmt);
             Exit();
+            Dispatcher.OnExitProgramStatement(stmt);
         }
 
         public virtual void OnAllocateStatement(AllocateStatement stmt)
```
Comment (fm-117) [inline TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs]: Dispatch calls are missing for OnAllocateStatement and OnFreeStatement

File: TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs [hunk 2]
```diff
@@ -1190,57 +1317,67 @@ public virtual void StartWhenConditionClause(List<TypeCobol.Compiler.CodeElement
...
         }
 
         public virtual void EndInvokeStatementConditional(TypeCobol.Compiler.CodeElements.InvokeStatementEnd end)
         {
             AttachEndIfExists(end);
             Exit();
+            Dispatcher.EndInvokeStatementConditional(end);
         }
 
         public virtual void StartJsonGenerateStatementConditional(TypeCobol.Compiler.CodeElements.JsonGenerateStatement stmt)
```
Comment (fm-117) [inline TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs]: Missing Dispatch calls for StartJsonGenerateStatementConditional and EndJsonGenerateStatementConditional

File: TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs [hunk 3]
```diff
@@ -5,6 +5,7 @@
 using TypeCobol.Compiler.CodeModel;
 using TypeCobol.Compiler.Nodes;
 using TypeCobol.Compiler.Parser;
+using Analytics;
```
Comment (fm-117) [inline TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs]: Not required ? (I haven't checked but according to the modifications it should not)

**Comment by smedilol** [comment]
Ok for me with last changes.

**Pull request merged**